pennylane/data/base/dataset.py

```diff
@@ -0,0 +1,272 @@
+from pennylane.data.base.attribute import AttributeInfo, AttributeType
+from pennylane.data.base.mapper import MapperMixin, match_obj_type
+from pennylane.data.base.typing_util import UNSET, T, ZarrAny, ZarrGroup
+
+
+@dataclass
+class Attribute(Generic[T]):
+    """
+    The Attribute class is used to declaratively define the
+    attributes of a Dataset subclass, in a similar way to
+    dataclasses. This class should not be used directly when
+    declaring attributes, use the ``attribute()`` function instead.
+
+    Attributes:
+        attribute_type: The ``AttributeType`` class for this attribute
+        info: Attribute info
+        default: Default value for this attribute, if not provided
+            to __init__()
+        default_factory: Zero-argument callable that returns
+            a default value
+    """
+
+    attribute_type: Type[AttributeType[ZarrAny, T, Any]]
+    info: AttributeInfo
+    default: Union[Literal[UNSET], T, None]
+    default_factory: Optional[Callable[[], T]] = None
+
+
+def attribute(  # pylint: disable=too-many-arguments, unused-argument
+    default: Union[Literal[UNSET], T, None] = UNSET,
+    attribute_type: Union[Type[AttributeType[ZarrAny, T, Any]], Literal[UNSET]] = UNSET,
+    doc: Optional[str] = None,
+    py_type: Optional[Any] = None,
+    extra: Optional[Dict[str, Any]] = None,
+    default_factory: Optional[Callable[[], T]] = None,
+    **kwargs,
+) -> Any:
+    """Used to define fields on a declarative Dataclass.
+
+    Args:
+        default: See ``Attribute.default``
+        attribute_type: ``AttributeType`` class for this attribute. If not provided,
+            type may be derived from the type annotation on the class.
+        doc: Documentation for the attribute
+        py_type: Type annotation or string describing this object's type. If not
+            provided, the annotation on the class will be used
+        default_factory: See ``Attribute.default_factory``
+    """
+
+    return Attribute(
+        cast(Type[AttributeType[ZarrAny, T, T]], attribute_type),
+        AttributeInfo(doc=doc, py_type=py_type, extra=extra or {}),
+        default=default,
+        default_factory=default_factory,
+    )
+
+
+@dataclass_transform(
+    order_default=False, eq_default=False, field_specifiers=(attribute,), kw_only_default=True
+)
+class _DatasetTransform:
+    """This base class that tells the type system that ``Dataset`` behaves like a dataclass.
+    See: https://peps.python.org/pep-0681/
+    """
+
+
+class Dataset(AttributeType[ZarrGroup, "Dataset", "Dataset"], MapperMixin, _DatasetTransform):
+    """
+    Base class for Datasets.
+
+    Attributes:
+        fields: A mapping of attribute names to their ``Attribute`` information. Note that
+            this contains attributes declared on the class, not attributes added to
+            an instance. Use ``attrs`` to view all attributes on an instance.
+        bind: The Zarr group that contains this dataset.
```

Comment:
Do we intend to be able to do something like `dataset = qml.data.Dataset(v1=1,bind='test_folder/data')` ? It didn't work for me.